java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Pattern;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S5917")
+public class DateTimeFormatterMismatchCheck extends IssuableSubscriptionVisitor {
+  private static final MethodMatchers OF_PATTERN_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.format.DateTimeFormatter")
+    .names("ofPattern")
+    .addParametersMatcher("java.lang.String")
+    .addParametersMatcher("java.lang.String", "java.util.Locale")
+    .build();
+
+  private static final MethodMatchers APPEND_VALUE_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.format.DateTimeFormatterBuilder")
+    .names("appendValue")
+    .addParametersMatcher("java.time.temporal.TemporalField", "int")
+    .build();
+
+  private static final MethodMatchers WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.temporal.WeekFields")
+    .names("weekBasedYear")
+    .addParametersMatcher()
+    .build();
+
+  private static final MethodMatchers WEEK_OF_WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.temporal.WeekFields")
+    .names("weekOfWeekBasedYear")
+    .addParametersMatcher()
```

Comment:
For clarity, I suggest to replace this with `addWithoutParametersMatcher`:
```suggestion
    .addWithoutParametersMatcher()
```